support/ttfautohint/frontend/ttfautohint.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.5.
.TH TTFAUTOHINT "1" "February 2012" "ttfautohint 0.7" "User Commands"
.SH NAME
ttfautohint \- add new auto-generated hints to a TrueType font
.SH SYNOPSIS
.B ttfautohint
[\fIOPTION\fR]... \fIIN-FILE OUT-FILE\fR
.br
.B ttfautohintGUI
[\fIOPTION\fR]...
.SH DESCRIPTION
Replace hints in TrueType font IN\-FILE and write output to OUT\-FILE.
The new hints are based on FreeType's autohinter.
.PP
These programs (for console and GUI, respectively)
are simple front\-ends to the `ttfautohint' library.
.PP
Long options can be given with one or two dashes,
and with and without equal sign between option and argument.
This means that the following forms are acceptable:
`\-foo=bar', `\-\-foo=bar', `\-foo bar', `\-\-foo bar'.
.SH OPTIONS
.TP
\fB\-f\fR, \fB\-\-latin\-fallback\fR
set fallback script to latin
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-\-help\-all\fR
show Qt and X11 specific options also
.TP
\fB\-i\fR, \fB\-\-ignore\-permissions\fR
override font license restrictions
.TP
\fB\-l\fR, \fB\-\-hinting\-range\-min\fR=\fIN\fR
the minimum ppem value for hint sets
.TP
\fB\-p\fR, \fB\-\-pre\-hinting\fR
apply original hints in advance
.TP
\fB\-r\fR, \fB\-\-hinting\-range\-max\fR=\fIN\fR
the maximum ppem value for hint sets
.TP
\fB\-v\fR, \fB\-\-verbose\fR
show progress information
.TP
\fB\-V\fR, \fB\-\-version\fR
print version information and exit
.TP
\fB\-x\fR, \fB\-\-x\-height\-snapping\-exceptions\fR=\fISTRING\fR
specify a comma\-separated list of
x\-height snapping exceptions
.SS "Qt Options:"
.TP
\fB\-\-graphicssystem\fR=\fISYSTEM\fR
select a different graphics system backend
instead of the default one
(possible values: `raster', `opengl')
.TP
\fB\-\-reverse\fR
set layout direction to right\-to\-left
.TP
\fB\-\-session\fR=\fIID\fR
restore the application for the given ID
.TP
\fB\-\-style\fR=\fISTYLE\fR
set application GUI style
(possible values: motif, windows, platinum)
.TP
\fB\-\-stylesheet\fR=\fISHEET\fR
apply the given Qt stylesheet
to the application widgets
.SS "X11 options:"
.TP
\fB\-\-background\fR=\fICOLOR\fR
set the default background color
and an application palette
(light and dark shades are calculated)
.TP
\fB\-\-bg\fR=\fICOLOR\fR
same as \fB\-\-background\fR
.TP
\fB\-\-btn\fR=\fICOLOR\fR
set the default button color
.TP
\fB\-\-button\fR=\fICOLOR\fR
same as \fB\-\-btn\fR
.TP
\fB\-\-cmap\fR
use a private color map on an 8\-bit display
.TP
\fB\-\-display\fR=\fINAME\fR
use the given X\-server display
.TP
\fB\-\-fg\fR=\fICOLOR\fR
set the default foreground color
.TP
\fB\-\-fn\fR=\fIFONTNAME\fR
set the application font
.TP
\fB\-\-font\fR=\fIFONTNAME\fR
same as \fB\-\-fn\fR
.TP
\fB\-\-foreground\fR=\fICOLOR\fR
same as \fB\-\-fg\fR
.TP
\fB\-\-geometry\fR=\fIGEOMETRY\fR
set the client geometry of first window
.TP
\fB\-\-im\fR=\fISERVER\fR
set the X Input Method (XIM) server
.TP
\fB\-\-inputstyle\fR=\fISTYLE\fR
set X Input Method input style
(possible values: onthespot, overthespot,
offthespot, root)
.TP
\fB\-\-name\fR=\fINAME\fR
set the application name
.TP
\fB\-\-ncols\fR=\fICOUNT\fR
limit the number of colors allocated
in the color cube on an 8\-bit display,
if the application is using the
QApplication::ManyColor color specification
.TP
\fB\-\-title\fR=\fITITLE\fR
set the application title (caption)
.TP
\fB\-\-visual\fR=\fIVISUAL\fR
force the application
to use the given visual on an 8\-bit display
(only possible value: TrueColor)
.PP
The programs accept both TTF and TTC files as input.
Use option \fB\-i\fR only if you have a legal permission to modify the font.
If option \fB\-f\fR is not set, glyphs not in the latin range stay unhinted.
The used ppem value for option \fB\-p\fR is FUnits per em, normally 2048.
.PP
A hint set contains the optimal hinting for a certain PPEM value;
the larger the hint set range, the more hint sets get computed,
usually increasing the output font size.  Note, however,
that the `gasp' table of the output file enables grayscale hinting
for all sizes.
.PP
If run in GUI mode, options not related to Qt or X11 set default values.
Additionally, there is no output to the console.
.PP
GUI support might be disabled at compile time.
.SH "REPORTING BUGS"
Report bugs to: freetype\-devel@nongnu.org
.br
FreeType home page: <http://www.freetype.org>
.SH COPYRIGHT
Copyright \(co 2011\-2012 Werner Lemberg <wl@gnu.org>.
License: FreeType License (FTL) or GNU GPLv2.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


support/ttfautohint/lib/ttfautohint.c
/* ttfautohint.c */

/*
 * Copyright (C) 2011-2012 by Werner Lemberg.
 *
 * This file is part of the ttfautohint library, and may only be used,
 * modified, and distributed under the terms given in `COPYING'.  By
 * continuing to use, modify, or distribute this file you indicate that you
 * have read `COPYING' and understand and accept it fully.
 *
 * The file `COPYING' mentioned in the previous paragraph is distributed
 * with the ttfautohint library.
 */


/* This file needs FreeType 2.4.5 or newer. */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "ta.h"


#define COMPARE(str) (len == (sizeof (str) - 1) \
                      && !strncmp(start, str, sizeof (str) - 1))


TA_Error
TTF_autohint(const char* options,
             ...)
{
  va_list ap;

  FONT* font;
  FT_Error error;
  FT_Long i;

  FILE* in_file = NULL;
  FILE* out_file = NULL;

  const char* in_buf = NULL;
  size_t in_len = 0;
  char** out_bufp = NULL;
  size_t* out_lenp = NULL;

  const unsigned char** error_stringp = NULL;

  FT_Long hinting_range_min = -1;
  FT_Long hinting_range_max = -1;

  TA_Progress_Func progress;
  void* progress_data;

  FT_Bool ignore_permissions = 0;
  FT_Bool pre_hinting = 0;
  FT_UInt fallback_script = 0;

  const char* op;


  if (!options || !*options)
  {
    error = FT_Err_Invalid_Argument;
    goto Err1;
  }

  /* XXX */
  va_start(ap, options);

  op = options;

  for(;;)
  {
    const char* start;
    size_t len;


    start = op;

    /* search comma */
    while (*op && *op != ',')
      op++;

    /* remove leading whitespace */
    while (isspace(*start))
      start++;

    /* check for empty option */
    if (start == op)
      goto End;

    len = op - start;

    /* the `COMPARE' macro uses `len' and `start' */

    /* handle option */
    if (COMPARE("error-string"))
      error_stringp = va_arg(ap, const unsigned char**);
    else if (COMPARE("fallback-script"))
      fallback_script = va_arg(ap, FT_UInt);
    else if (COMPARE("hinting-range-max"))
      hinting_range_max = (FT_Long)va_arg(ap, FT_UInt);
    else if (COMPARE("hinting-range-min"))
      hinting_range_min = (FT_Long)va_arg(ap, FT_UInt);
    else if (COMPARE("ignore-permissions"))
      ignore_permissions = (FT_Bool)va_arg(ap, FT_Int);
    else if (COMPARE("in-buffer"))
    {
      in_file = NULL;
      in_buf = va_arg(ap, const char*);
    }
    else if (COMPARE("in-buffer-len"))
    {
      in_file = NULL;
      in_len = va_arg(ap, size_t);
    }
    else if (COMPARE("in-file"))
    {
      in_file = va_arg(ap, FILE*);
      in_buf = NULL;
      in_len = 0;
    }
    else if (COMPARE("out-buffer"))
    {
      out_file = NULL;
      out_bufp = va_arg(ap, char**);
    }
    else if (COMPARE("out-buffer-len"))
    {
      out_file = NULL;
      out_lenp = va_arg(ap, size_t*);
    }
    else if (COMPARE("out-file"))
    {
      out_file = va_arg(ap, FILE*);
      out_bufp = NULL;
      out_lenp = NULL;
    }
    else if (COMPARE("pre-hinting"))
      pre_hinting = (FT_Bool)va_arg(ap, FT_Int);
    else if (COMPARE("progress-callback"))
      progress = va_arg(ap, TA_Progress_Func);
    else if (COMPARE("progress-callback-data"))
      progress_data = va_arg(ap, void*);

    /*
      x-height-snapping-exceptions
     */

  End:
    if (!*op)
      break;
    op++;
  }

  va_end(ap);

  /* check options */

  if (!(in_file
        || (in_buf && in_len)))
  {
    error = FT_Err_Invalid_Argument;
    goto Err1;
  }

  if (!(out_file
        || (out_bufp && out_lenp)))
  {
    error = FT_Err_Invalid_Argument;
    goto Err1;
  }

  font = (FONT*)calloc(1, sizeof (FONT));
  if (!font)
  {
    error = FT_Err_Out_Of_Memory;
    goto Err1;
  }

  if (hinting_range_min >= 0 && hinting_range_min < 2)
  {
    error = FT_Err_Invalid_Argument;
    goto Err1;
  }
  if (hinting_range_min < 0)
    hinting_range_min = 8;

  if (hinting_range_max >= 0 && hinting_range_max < hinting_range_min)
  {
    error = FT_Err_Invalid_Argument;
    goto Err1;
  }
  if (hinting_range_max < 0)
    hinting_range_max = 1000;

  font->hinting_range_min = (FT_UInt)hinting_range_min;
  font->hinting_range_max = (FT_UInt)hinting_range_max;

  font->progress = progress;
  font->progress_data = progress_data;

  font->ignore_permissions = ignore_permissions;
  font->pre_hinting = pre_hinting;
  /* restrict value to two bits */
  font->fallback_script = fallback_script & 3;

  /* now start with processing the data */

  if (in_file)
  {
    error = TA_font_file_read(font, in_file);
    if (error)
      goto Err;
  }
  else
  {
    /* a valid TTF can never be that small */
    if (in_len < 100)
    {
      error = FT_Err_Invalid_Argument;
      goto Err1;
    }
    font->in_buf = (FT_Byte*)in_buf;
    font->in_len = in_len;
  }

  error = TA_font_init(font);
  if (error)
    goto Err;

  /* loop over subfonts */
  for (i = 0; i < font->num_sfnts; i++)
  {
    SFNT* sfnt = &font->sfnts[i];


    error = FT_New_Memory_Face(font->lib, font->in_buf, font->in_len,
                               i, &sfnt->face);
    if (error)
      goto Err;

    error = TA_sfnt_split_into_SFNT_tables(sfnt, font);
    if (error)
      goto Err;

    if (font->pre_hinting)
      error = TA_sfnt_create_glyf_data(sfnt, font);
    else
      error = TA_sfnt_split_glyf_table(sfnt, font);
    if (error)
      goto Err;

    /* check permission */
    if (sfnt->OS2_idx != MISSING)
    {
      SFNT_Table* OS2_table = &font->tables[sfnt->OS2_idx];


      /* check lower byte of the `fsType' field */
      if (OS2_table->buf[OS2_FSTYPE_OFFSET + 1] == 0x02
          && !font->ignore_permissions)
      {
        error = TA_Err_Missing_Legal_Permission;
        goto Err;
      }
    }
  }

  /* build `gasp' table */
  error = TA_sfnt_build_gasp_table(&font->sfnts[0], font);
  if (error)
    goto Err;

  /* XXX handle subfonts for bytecode tables */

  /* build `cvt ' table */
  error = TA_sfnt_build_cvt_table(&font->sfnts[0], font);
  if (error)
    goto Err;

  /* build `fpgm' table */
  error = TA_sfnt_build_fpgm_table(&font->sfnts[0], font);
  if (error)
    goto Err;

  /* build `prep' table */
  error = TA_sfnt_build_prep_table(&font->sfnts[0], font);
  if (error)
    goto Err;

  /* hint the glyphs and build bytecode */
  error = TA_sfnt_build_glyf_hints(&font->sfnts[0], font);
  if (error)
    goto Err;

  /* loop again over subfonts */
  for (i = 0; i < font->num_sfnts; i++)
  {
    SFNT* sfnt = &font->sfnts[i];


    error = TA_sfnt_build_glyf_table(sfnt, font);
    if (error)
      goto Err;
    error = TA_sfnt_build_loca_table(sfnt, font);
    if (error)
      goto Err;
    error = TA_sfnt_update_maxp_table(sfnt, font);
    if (error)
      goto Err;

    /* we add one glyph for composites */
    if (sfnt->max_components && !font->pre_hinting)
    {
      error = TA_sfnt_update_hmtx_table(sfnt, font);
      if (error)
        goto Err;
      error = TA_sfnt_update_post_table(sfnt, font);
      if (error)
        goto Err;
      error = TA_sfnt_update_GPOS_table(sfnt, font);
      if (error)
        goto Err;
    }
  }

  if (font->num_sfnts == 1)
    error = TA_font_build_TTF(font);
  else
    error = TA_font_build_TTC(font);
  if (error)
    goto Err;

  if (out_file)
  {
    error = TA_font_file_write(font, out_file);
    if (error)
      goto Err;
  }
  else
  {
    *out_bufp = (char*)font->out_buf;
    *out_lenp = font->out_len;
  }

  error = TA_Err_Ok;

Err:
  TA_font_unload(font, in_buf, out_bufp);

Err1:
  if (error_stringp)
    *error_stringp = (const unsigned char*)TA_get_error_message(error);

  return error;
}

/* end of ttfautohint.c */
